File: distance1.py
# Copyright (c) 2019 Alexander Lopatin. All rights reserved.
import time
import sys


def fee(c): return ord(c)


def distance1(a, b):  # Algorithm #1 (Reference): Brute force.
    if a == b:  # success, if strings are equal
        return 0
    d = sys.maxsize  # Enough for the worst sum of 8421504 bytes in 32bit.
    for i, c in enumerate(a):  # for each index and character in  a
        d = min(d, fee(c) + distance1(a[:i]+a[i+1:], b))
    for i, c in enumerate(b):
        d = min(d, fee(c) + distance1(a, b[:i]+b[i+1:]))
    return d  # minimal achieved distance for a & b strings


if __name__ == '__main__':
    assert distance1("cool", "cold") == fee('o') + fee('d')
    assert distance1("!~!", "~!!") == 2 * fee('!')  # ! vs ~ (33 vs 126)
    t = time.process_time()
    n = 0  # cycle number
    dt = 0  # distance total sum
    while n < 100000:
        s = "{0:05d}".format(n)  # 00000, 00001, 00002, ..., 99999
        for j in range(6):
            e = s[:j]  # "", "0", "00", "000", "0000", "00000" ...
            f = s[j:]  # "00000", "0000", "000", "00", "0", "" ...
            de = distance1(e, f)
            df = distance1(f, e)
            dt += de
            assert(de == df)
        n += 1
    n = n * 12  # now it is number of distance1 calls
    t = time.process_time() - t
    print(t, dt, n, t/n, sep=',')

File: distance1.swift
//  Copyright (c) 2019 Alexander Lopatin. All rights reserved.
import Foundation

func fee(_ c: Character) -> Int {
    var i = 255
    if let v = c.unicodeScalars.first?.value {
        i = Int(v)
        if i < 0 || i > 255 {
            i = 255
        }
    }
    return i
}

// Algorithm #1: Brute force.
func distance1a(_ a: inout [UInt8], _ b: inout [UInt8]) -> Int {
    if a == b {
        return 0
    }
    var d = Int.max
    for i in a.indices {
        let n = a.remove(at: i)
        d = min(d, Int(n) + distance1a(&a , &b))
        a.insert(n, at: i)
    }
    for i in b.indices {
        let n = b.remove(at: i)
        d = min(d, Int(n) + distance1a(&a , &b))
        b.insert(n, at: i)
    }
    return d
}

extension String {
    func byteArray() -> [UInt8] {
        var a = [UInt8]()
        for c in self.unicodeScalars {
            a.append(c.value < UInt8.max ? UInt8(c.value) :UInt8.max)
        }
        return a
    }
}

func distance1(_ a: String, _ b: String) -> Int {
    var aA = a.byteArray()
    var bA = b.byteArray()
    return distance1a(&aA, &bA)
}

assert(distance1("at", "cat") == fee("c"))
assert(distance1("cool", "cold") == fee("o") + fee("d"))
assert(distance1("!~!", "~!!") == 2 * fee("!")) // ! vs ~ (33 vs 126)

let date = Date()
var d = 0
var n = 0
while n < 100000 {

    func test(_ a: String, _ b: String) {
        let da = distance1(a, b)
        let db = distance1(b, a)
        d += da
        assert(da == db)
    }

    let s = String(format: "%05d", n)
    test("", s)
    for i in s.indices {
        test(String(s[...i]), String(s[s.index(after: i)...]))
    }
    n += 1
}
n *= 12
let t = Date().timeIntervalSince(date as Date)
print("\(t),\(d),\(n),\(t.magnitude/Double(n))")
